src/realm/object-store/sync/impl/emscripten/socket_provider.cpp

```diff
@@ -0,0 +1,219 @@
+
+    static EM_BOOL message_callback(int, const EmscriptenWebSocketMessageEvent* event, void* user_data)
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        REALM_ASSERT(!event->isText);
+        observer->websocket_binary_message_received(
+            util::Span(reinterpret_cast<char*>(event->data), event->numBytes));
+        return EM_TRUE;
+    }
+
+    static EM_BOOL error_callback(int, const EmscriptenWebSocketErrorEvent*, void* user_data)
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        observer->websocket_error_handler();
+        return EM_TRUE;
+    }
+
+    static EM_BOOL close_callback(int, const EmscriptenWebSocketCloseEvent* event, void* user_data)
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        REALM_ASSERT(event->code >= 1000 && event->code < 5000);
+        auto status = event->code == 1000 ? Status::OK() : Status(ErrorCodes::Error(event->code), event->reason);
+        observer->websocket_closed_handler(event->wasClean, std::move(status));
+        return EM_TRUE;
+    }
+
+    EMSCRIPTEN_WEBSOCKET_T m_socket;
+    std::shared_ptr<LivenessSentinel> m_sentinel;
+    std::unique_ptr<WebSocketObserver> m_observer;
+};
+} // namespace
+
+EmscriptenSocketProvider::EmscriptenSocketProvider()
+    : SyncSocketProvider()
+    , EmscriptenScheduler()
+{
+}
+
+EmscriptenSocketProvider::~EmscriptenSocketProvider() = default;
+
+SyncSocketProvider::SyncTimer EmscriptenSocketProvider::create_timer(std::chrono::milliseconds delay,
+                                                                     FunctionHandler&& handler)
+{
+    util::EmscriptenScheduler& scheduler = *this;
+    return std::make_unique<EmscriptenTimer>(delay, std::move(handler), scheduler);
+}
+
+std::unique_ptr<WebSocketInterface> EmscriptenSocketProvider::connect(std::unique_ptr<WebSocketObserver> observer,
+                                                                      WebSocketEndpoint&& endpoint)
+{
+    // Convert the list of protocols to a string
+    std::ostringstream protocol_list;
+    protocol_list.exceptions(std::ios_base::failbit | std::ios_base::badbit);
+    protocol_list.imbue(std::locale::classic());
+    if (endpoint.protocols.size() > 1)
+        std::copy(endpoint.protocols.begin(), endpoint.protocols.end() - 1,
+                  std::ostream_iterator<std::string>(protocol_list, ","));
+    protocol_list << endpoint.protocols.back();
+    std::string protocols = protocol_list.str();
+    for (size_t i = 0; i < protocols.size(); i++) {
+        if (protocols[i] == '/')
+            protocols[i] = '#';
+    }
+
+    std::string url =
+        util::format("%1://%2:%3%4", endpoint.is_ssl ? "wss" : "ws", endpoint.address, endpoint.port, endpoint.path);
+
+    EmscriptenWebSocketCreateAttributes attr;
+    emscripten_websocket_init_create_attributes(&attr);
+    attr.protocols = protocols.c_str();
+    attr.url = url.c_str();
```

Comment:
I think this is the opposite of what we want. We always need the websocket to be bound to the current thread, because if we were to run the sync client inside a worker `createOnMainThread` will always proxy the websocket calls from the worker to the main thread. The header says `If the created WebSocket only needs to be accessible on the thread that created it, and the creating thread is an event based thread (meaning it regularly yields back to the browser event loop), then it is more efficient to set this to false.` which I think is exactly our case.

Reply:
> I think this is the opposite of what we want. 

Sort of. I don't think we really care if the websocket is usable from other threads, even if we never use it from other threads. If I understand the documentation correctly, setting this to true is the *safer* option and is _correct_ in all cases, even if you _could_ use false. The only benefit to _false_ seems to be performance (although there may also be some implicit cleanup if your thread exits, but we hopefully don't need that). That was why I suggested setting it to true when we first do multi-threading so that we can deal with other issues that are sure to pop up before we need to prove that this is safe.

At the very least I think we should set it explicitly rather than using the default.

Reply:
I agree, setting it explicitly to false makes sense. My argument is against ever setting it to true. But we are still some ways away from multi-threaded WebAssembly anyway, so we can revisit this at a later date.